keps/sig-network/4427-relaxed-dns-search-validation/README.md

```diff
@@ -0,0 +1,813 @@
+within Kubernetes. The DNS searches in the Kubernetes pods spec are not defining the host name, so even under the stricter interpretation
+of RFC-1123, relaxing these checks does not allow users of Kubernetes to introduce an invalid host name where one did not
+already exist outside of Kubernetes.
+
+### Goals
+
+- Support workloads that need to resolve DNS short names where the full DNS name includes characters disallowed by RFC-1123.
+
+## Proposal
+
+Introduce a RelaxedDNSSearchValidation feature gate which is disabled by default. When the feature gate is enabled,
+DNS search strings will be validated according to RFC-2181:
+- The length of any one label is limited to between 1 and 63 octets
+- A full domain name is limited to 255 octets (including the separators)
+
+Since the relaxed check allows previously invalid values, care must be taken to support cluster downgrades safely. To accomplish this, the validation will distinguish between new resources and updates to existing resources:
+-	When the feature gate is disabled:
+  - (a) New resources will use strict validation based on RFC-1123 (no change to current validation)
+  - (b) Updates to existing resources will use relaxed validation if any search string in the existing list fails strict validation
+- When the feature gate is enabled:
+  - (c) New resources will use relaxed validation.
+  - (d) Updates to existing resources will use relaxed validation.
+
+This means that it is safe to downgrade a cluster with the feature gate enabled to a version where the feature gate is present (whether it’s enabled or disabled). It is not safe, in general, to downgrade from a cluster with the gate enabled to a version prior to the gate being introduced, since values may have been written to storage which will no longer pass validation. However, this scenario requires opting in through enabling the gate. In practice, the recommended approach would be to only enable to the gate after upgrading from a version with relaxed checking already present.
+
+As long as the gate is disabled, there is no compatibility change, so cluster downgrades are not affected by the feature.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Test Plan
+
+<!--
```

Comment:
I think this just needs to describe that we will cover the validation logic in unit tests and ...do we need e2e coverage of this?

Reply:
+1

I noticed that rollback is described above. Just so it's clear that this part of PRR, refer to it from here or copy-paste the relevant text to here?

Reply:
Ah Nevermind, this is just the section header. I got ahead of myself.